hibernate-core/src/main/java/org/hibernate/metamodel/model/domain/internal/JpaMetamodelImpl.java

```diff
@@ -319,7 +319,22 @@ public Set<EmbeddableType<?>> getEmbeddables() {
 
 	@Override
 	public EnumJavaType<?> getEnumType(String className) {
-		return enumJavaTypes.get( className );
+		if ( enumJavaTypes.containsKey( className ) ) {
+			return enumJavaTypes.get( className );
+		}
+		final ClassLoaderService classLoaderService = serviceRegistry.getService( ClassLoaderService.class );
+		try {
+			final Class<Object> clazz = classLoaderService.classForName( className );
+			if ( clazz == null || !clazz.isEnum() ) {
+				return null;
+			}
+			final EnumJavaType enumType = new EnumJavaType( clazz );
+			enumJavaTypes.put( className, enumType );
+			return enumType;
```

Comment:
Mutating the map here is a concurrency issue, so let's not do that.

```suggestion
			return new EnumJavaType( clazz );
```

Reply:
Right, I wanted to avoid creating new object each time, but concurrency is more serious problem.